### ISSUE TITLE
Uploading AIP to Swift

### ISSUE THREAD

**Issue by murny** [ualbertalib/pushmi_pullyu#26]
PUT request for uploading AIP to Swift using it's REST API with checksum of bag (using swift api auth token).

No idea how Swift works, so requires some investigation with using it's API for uploading files. Use a ruby http client API to do these requests. Try to use the builtin Ruby libraries preferably. 

Swift REST API documentation: https://developer.openstack.org/api-ref/object-storage/
Chunking of large files (>5gb): https://docs.openstack.org/developer/swift/api/large_objects.html

https://github.com/ruby-openstack/ruby-openstack is probably best Gem out there for this. It also supports etag/chunking of large files as well.

Example script: https://gist.github.com/murny/c4696ec251ec43214303d218d525837b

**Comment by murny** [comment]
Any updates on this @astrilet?  Need help with anything?

**Comment by murny** [comment]
@astrilet can you update this ticket please? 

Maybe create a branch and push your code up so people can see where you are at with this and provide help if you need it? 

You think you will be finished this by early next week? (monday/tuesday)?

**Comment by murny** [comment]
Hey @astrilet, before I can close this and consider this done. I need some documentation on how we would support uploading 5 GB+ files to swift. I remember in standups you mention playing around with this. Can you add some documentation to this ticket on how this would look like if we were to implement this?

I believe `ruby-openstack gem` has support for this via chunking data steams and supplying a manifests. 

Thanks.

**Comment by astrilet** [comment]
swift has a limit of upload file size 5 GB. If you have file larger then 5 GB it needs to be broken into smaller pieces. Python client, for example, does it for you. If we need to upload files > 5GB we need to implement logic to break file into chunks as well. I had some code to do that  but as I understood we decided not to implement it right now and leave it as a future release feature.

**Comment by murny** [comment]
@astrilet would you be able to add this code of yours that you have into this ticket? Or maybe create a gist with it?

This will help everyone for when we want to implement this functionality in the future. Thanks.

**Comment by astrilet** [comment]
OK will create gist and share some code

**Comment by astrilet** [comment]
<script src="https://gist.github.com/astrilet/b383667de73ffb3c7560e53f24411583.js"></script>

**Comment by astrilet** [comment]
Code was committed in PR#57

**Comment by murny** [comment]
Can you make both code snippets into a single gist? and If dynamic is the preferred way lets just use that way (aka remove anything relating to the static manifest code stuff).

Clean up the gist too, you have a ton of commented out code and a ton of duplication of code.

 Basically you should have two paths of your code...

Given a file of 5GB or lower...it should uses the existing code and simply just upload the file into swift.

Given a file of 5GB+ (lets say its 13 GBs), it should get chunked and uploaded and then a manifest gets uploaded for the chunks.

Also right now, your `split` is splitting the file in 1MB chunks...so that 13GB file gets seg001 to seg999 before it crashes (as it ran out of suffixes).

**Comment by astrilet** [comment]
sorry static method is proffered

**Comment by astrilet** [comment]
This is not a production code, just proof of concept and discovery how it works. Static file segments have to be at least 1 MB in size. I will write some discovery comments.

**Comment by astrilet** [comment]
I will create separate file with code more suitable for production use.

**Comment by murny** [comment]
Here's a list for this to be completed:
- [ ]  A example implementation of dynamic manifest way (in a gist)
- [ ]  A example implementation of static manifest way (in a seperate gist)
- [ ] Both should be implemented in a "production" way. Aka clean, concised code and using the same current API entry point (deposit_file() method)
 - [ ] A write up of static vs dynamic. Pros and cons of each. Why static is apparently better? ( Apparently it can do checksums while dynamic can not, is this correct?). Include links to documentation you found (official or non official) for this stuff.
 - [ ] Is there any issues with 5GB+ files with our current `openstack ruby` gem? Could this be motivation to just use raw http rest calls instead?  Or can openstack Ruby gem do everything we need to do here for this?
- [ ] Document other issues around manifests and segmented files in Swift. If I upload a big file and I delete one of the segments what happens when I download the main object? Does it do a integrity check? If I upload a giant file (in segments) and then later I reduce the file size and update it it swift, what happens? Does it handle cleaning out the previous old segments and replace them when I update it? Etc.
- [ ] Also you discovered a bug in how we do updating and creating a new swift object (creating a new object actually uploads twice, a file with zero content and then updates this with the actual file). Can you create an github issue in the PushmiPullyu repo for this so this is tracked somewhere?

**Comment by weiweishi** [comment]
Some discoveries from the Python client in our Swift instance:

- It uses the dynamic large object (DLO)
- The manifest object (master record) lives in the main_container and have the object_manifest metadata pointing to the right container. The full list of metadata is listed as following:
```    Account: *****
     Container: test
        Object: ditest/1.mov
  Content Type: video/quicktime
Content Length: 13732725786
 Last Modified: Wed, 07 Jun 2017 22:54:14 GMT
          ETag: "ebe1be2ac9a8b7b2d519382e80d70210"
      Manifest: test_segments/ditest/1.mov/1444374367.000000/14806467610/1073741824/
    Meta Mtime: 1444374367.000000
 Accept-Ranges: bytes
   X-Timestamp: 1496876053.55277
    X-Trans-Id: tx9ceae802478742cd9e2c0-0059397716
``` 
- it calculates a checksum based on all the segments uploaded. (not sure what algorithm Swift uses, as it doesn't match any checksum result generated on my workstation, with md5sum, sha1sum, and sha256sum, sha3sum)

- It automatically created a separate segment container, called main_container_segments (if the targeted container for uploading is main_container)

- It then uses a Mtime + some string not sure what is based on + segment sequenced id as the path for the segments in the following format:
ditest/1.mov/1444374367.000000/14806467610/1073741824/00000001

- I removed a segment and it resulted in a download error. 

- I uploaded a new version but it saved to the same segment path and I can't seem to find the old version. There's no "Archive" container etc, and no X-History-Location or X-Versions-Location. The new version seems to have replaced the old version so it's an issue
  - Does versioning properly turn on in our Swift environment?
  - How does the versioning handled for DLO in versions. Are they moved together to a different location?

**Comment by astrilet** [comment]
Some investigation on how swift handles large files.
Swift objects have size limit of 5GB. Large objects are handled by splitting them into segments and uploading each segment as individual file. In addition manifest object is created with links all segments together. There are two types of handling large objects using static or dynamic large objects. 
Differences (taken form https://docs.openstack.org/user-guide/cli-swift-large-object-creation.html)
End-to-end integrity: static - assured, dynamic - not guaranteed
Manifest file: static - you have to create your own manifest, dynamic - created automatically for you.
Upload order: static - first upload all segments then manifest, dynamic - any order
Removal or addition of segments: static - can not be done, dynamic - you can add or remove segments
Segment size: static min 1 MB, dynamic - any size
Location: static - can be in any container, dynamic - all segments have to be in the same container (Please note that manifest does not have to be in the same container as segments).
Our biggest concern with using dynamic large objects is that if one or more segments are deleted form the container (for whatever reason) object would still be downloadable without any errors.

**Comment by astrilet** [comment]
For versioning to be working on the container that you created you need to instruct swift to archive overwritten files into the some other container (archive storage container) 
See the link:
https://docs.openstack.org/user-guide/cli-swift-set-object-versions.html

**Comment by astrilet** [comment]
I looks like in case of swift python client all segments are stored in different container with unique id that is derived form file type and file size. If you upload new version of a file, it will most likely be of a different size and therefore will have different name, so will not override existing segments. However I have used swift python client to upload 4 segments of a large file into a swift, then went into swift and deleted one segment (last one) and was still able do download file without an error, it was just missing last segment and was a smaller size.

**Comment by astrilet** [comment]
I have updated my code and created separate gist for splitting large files into smaller chunks and creating dynamic manifest. Link to my gist:
https://gist.github.com/astrilet/03c3bbf6a9b68b9343bbd5ed06958768

**Comment by astrilet** [comment]
Here is a link to gist that creates static large object. 
https://gist.github.com/astrilet/94a858221c6aea19390c7403647f2331

**Comment by astrilet** [comment]
Apparently there is a concern with using dynamic large objects. Manifest of dynamic large object just provides pattern of all parts but does not know how many parts are should be there. Swift will just concatenate all parts that match given pattern into single file during download. There if once file segment is deleted, swift will still download file without error. Static dynamic object do not have such behaviour since manifest specifies exact location of all parts, their sizes and md5 checksum. Usage of static large object therefore is preferred.

**Comment by astrilet** [comment]
Ruby openstack gem does not have any limitation build in for 5GB+ file size, limitation comes form SWIFT itself, it can not store objects larger then 5GB, therefore it is not a openstack gem limitation, it is SWIFT limitation.

**Comment by astrilet** [comment]
In the current python implementation of swift client as well as in my sample gist code segments are stored int the different container so that it does not clutter main container with segments. Segments are stored in the {main_container_name}_segments container. usually segments name include some unique file string (file size/last 5 digits form md5 checksum/download datetime, etc}  plus segment id to make segment of different version of the same file unique. Therefore we do not have to worry about overwriting segments when inserting new version of the file. Main manifest file is stored in the main container and  is versioned. As a result data of all versions are preserved.

**Comment by astrilet** [comment]
Some investigation on how swift handles large files.

Ruby openstack gem does not have any limitation build in for 5GB+ file size, limitation comes form SWIFT itself, it can not store objects larger then 5GB, therefore it is not a openstack gem limitation, it is SWIFT limitation.

Swift objects have size limit of 5GB. Large objects are handled by splitting them into segments and uploading each segment as individual file. In addition manifest object is created with links all segments together. There are two types of handling large objects using static or dynamic large objects.
Differences (taken form https://docs.openstack.org/user-guide/cli-swift-large-object-creation.html)
End-to-end integrity: static - assured, dynamic - not guaranteed
Manifest file: static - you have to create your own manifest, dynamic - created automatically for you.
Upload order: static - first upload all segments then manifest, dynamic - any order
Removal or addition of segments: static - can not be done, dynamic - you can add or remove segments
Segment size: static min 1 MB, dynamic - any size
Location: static - can be in any container, dynamic - all segments have to be in the same container (Please note that manifest does not have to be in the same container as segments).
Our biggest concern with using dynamic large objects is that if one or more segments are deleted form the container (for whatever reason) object would still be downloadable without any errors.

**Comment by weiweishi** [comment]
@astrilet I've edited your answer with proper markup to make the text more readable. But your answer doesn't seem to be completed. I'm also wondering about the question of what the Openstack gem can offer us that plain curl command might not be able to offer us easily.

**Comment by astrilet** [comment]
updated my comments

**Comment by murny** [comment]
I think there is some confusion here with regards to the points on the openstack ruby gem.  I already know the file size limitation is on Swift side and not a limitation with openstack ruby gem. And I know that using ruby is a much preferred way of doing this work over shell commands and other things.

My original question was:

> Is there any issues with 5GB+ files with our current openstack ruby gem? Could this be motivation to just use raw http rest calls instead? Or can openstack Ruby gem do everything we need to do here for this?

By this I mean, was there any pain points implementing static/dynamic manifest workflows with openstack ruby gem? Would it be easier to do this in REST calls using [ruby's http client library](https://ruby-doc.org/stdlib-2.4.1/libdoc/net/http/rdoc/Net/HTTP.html) (which is basically what the gem is doing itself. Except it is adding a bunch of stuff we don't care about or doing things maybe not in the best of ways)? Or was everything easily doable here with the gem?

We have discovered other pain points with this gem, so was wondering if this (the implementation of handling big files) could be yet another pain point which might make us reconsider this gem?

So basically looking for a yes or no answer to my question and if there was any gotchas we should be aware of when you were implementing your gists with the gem?

**Comment by astrilet** [comment]
I do not think there is more "pain points" implementing handling large files with current openstack gem than using raw http requests. One way or another we need to handle large files in a special way by breaking them into smaller chunks, uploading each chunk individually and creating manifest file. The only way to bypass that is to use swift python client, but then it creates dynamic large objects which is probably not what we want. 
For now we have working solution and handling large files does not seem to be any more "painful" then using http requests. In my opinion we should use the existing solution and just extend it to be able to handle large files using code examples in my gists.

**Comment by weiweishi** [comment]
uploading aip to swift is completed, with the use of the [openstack ruby gem](https://github.com/ruby-openstack/ruby-openstack). 

This function will be handled in https://github.com/ualbertalib/swift_ingest, so that we can extend the use of the new swift_ingest gem to other ingest workflows.
